src/action/config/ActionConfiguration.cpp

```diff
@@ -225,23 +225,22 @@ void ActionConfiguration::createAction()
 
   // Determine data and mesh
   std::vector<int> sourceDataIDs;
-  int              targetDataID = -1;
-  mesh::PtrMesh    mesh;
-  for (const mesh::PtrMesh &aMesh : _meshConfig->meshes()) {
-    if (aMesh->getName() == _configuredAction.mesh) {
-      mesh = aMesh;
-      for (const mesh::PtrData &data : mesh->data()) {
-        if (std::find(_configuredAction.sourceDataVector.begin(), _configuredAction.sourceDataVector.end(), data->getName()) != _configuredAction.sourceDataVector.end()) {
-          sourceDataIDs.push_back(data->getID());
-        }
-        if (data->getName() == _configuredAction.targetData) {
-          targetDataID = data->getID();
-        }
-      }
+  int              targetDataID;
+  PRECICE_CHECK(_meshConfig->hasMeshName(_configuredAction.mesh),
+                "Data action uses mesh \"{}\" which is not configured. Please ensure that the correct mesh name is given in <action:python mesh=\"...\">", _configuredAction.mesh);
+  mesh::PtrMesh mesh = _meshConfig->getMesh(_configuredAction.mesh);
+
+  if (mesh->hasDataName(_configuredAction.targetData)) {
+    targetDataID = mesh->data(_configuredAction.targetData)->getID();
+  } else {
+    targetDataID = -1;
+  }
+
+  for (const mesh::PtrData &data : mesh->data()) {
+    if (std::find(_configuredAction.sourceDataVector.begin(), _configuredAction.sourceDataVector.end(), data->getName()) != _configuredAction.sourceDataVector.end()) {
+      sourceDataIDs.push_back(data->getID());
```

Comment:
Not sure whether we should actually touch this code, at all. I don't really see a clear improvement here. @fsimonis any opinion/idea?

Reply:
This actually looks buggy as the sourceData is now in the order of the mesh->data and not how it was originally declared.
You could simplify (and probably fix) this by iterating over the `_configuredAction.sourceDataVector` and check if the `mesh` has this data. Then we know that all dataIDs are correct after the loop.

Reply:
I hope I got you right. Tried this in 2c203a9b661d537943c0db863af96eb5fc48bdc0.